docs/components/index.md

```diff
@@ -0,0 +1,360 @@
+
+## Use of Razor
+
+Razor is a syntax for embedding C# code into HTML markup files. Files containing Razor generally have a *\*.cshtml* file extension. We recommend that you become familiar with Razor while working with Blazor. For more information, see the [Razor syntax reference](https://docs.microsoft.com/aspnet/core/mvc/views/razor). Note that not all of the features of Razor are available in Blazor at this time.
+
+The following pair of Blazor components demonstrates how Blazor takes advantage of features in Razor, C#, and HTML. Many of these features are described in more detail later in this topic.
+
+Index page component (*Index.cshtml*):
+
+[!code-cshtml[](common/samples/2.x/ComponentsSample/Pages/Index.cshtml?start=1&end=11)]
+
+When the index page loads, the `HeadingComponent` loads. The component:
+
+* Displays a heading on the page in title case.
+* Presents a form that allows the user to control the heading's font style.
+
+Heading component (*HeadingComponent.cshtml*):
+
+[!code-cshtml[](common/samples/2.x/ComponentsSample/Pages/HeadingComponent.cshtml)]
+
+## Razor directives
+
+A Razor file specifies [Razor directives](https://docs.microsoft.com/aspnet/core/mvc/views/razor#directives) at the top of the file. Razor directives supported by Blazor are shown in the table. `@model`, `@section`, and Tag Helper directives aren't supported at this time.
+
+| Directive | Description |
+| --------- | ----------- |
+| [@functions](https://docs.microsoft.com/aspnet/core/mvc/views/razor#functions) | Adds a C# code block to a component. |
+| `@implements` | Implements an interface for the generated component class. Often used to create a component layout by implementing `ILayoutComponent`. |
+| [@inherits](https://docs.microsoft.com/aspnet/core/mvc/views/razor#inherits) | Provides full control of the class that the component inherits. |
+| [@inject](https://docs.microsoft.com/aspnet/core/mvc/views/razor#inject) | Enables service injection from the [service container](https://docs.microsoft.com/aspnet/core/fundamentals/dependency-injection). For more information, see [Dependency injection into views](https://docs.microsoft.com/aspnet/core/mvc/views/dependency-injection). |
+| `@layout` | Specifies a layout component. Layout components are used to avoid code duplication and inconsistency. |
+| [@page](https://docs.microsoft.com/aspnet/core/mvc/razor-pages#razor-pages) | Specifies that the component should handle requests directly. The `@page` directive can be specified with a route and optional parameters. Unlike Razor Pages, the `@page` directive doesn't need to be the first directive at the top of the file. |
+| [@using](https://docs.microsoft.com/aspnet/core/mvc/views/razor#using) | Adds the C# `using` directive to the generated component class. |
+
+## Component classes
```

Comment:
I like the heading for this section, but the first paragraph doesn't say anything about how Blazor components are implemented as .NET classes and it seems to go out of its way to immediately point out something that you can't do easily today.

I'd expect this section to cover:
- Components are implemented using classes
- Components are typically implemented in .cshtml files using a combination of HTML and C#
- Dynamic rendering logic (loops, conditionals, expressions, etc.) is added using an embedded C# syntax called Razor.
- The HTML markup and C# rendering logic are converted into a component class at build time. 
- The name of the generated .NET class matches the name of the file.
- Members of the component class are defined in a `@functions` block. In the `@functions` block you can specify component state (properties, fields) as well as methods for event handling or for defining other component logic.
- Component members can then be used as part of the component's rendering logic using C# expressions that start with `@` (partially covered by the Render C# expressions section already)

Reply:
Agreed.

Can we remove the statements about what *isn't supported at this time*? It suggests we plan to add those things later, which is not the case for things we don't think fit into the programming model.

Reply:
I think we should move this section to the top of the topic and let the much-shorter *BlazorRocks* example be the first example the readers see. The larger example of how to change the heading style is still available in the sample and still used in the code snippets below. On the next commit, let's take a look and see if it works better.

Reply:
I would prefer we predominantly show components implemented inline and not using the separate base class approach. We may eventually decide to support a partial class model, but for now the recommended practice is to implement components inline in a single .cshtml file.